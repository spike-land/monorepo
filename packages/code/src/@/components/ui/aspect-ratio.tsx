import * as AspectRatioPrimitive from "@radix-ui/react-aspect-ratio";

const AspectRatio = AspectRatioPrimitive
  .Root satisfies typeof AspectRatioPrimitive.Root as typeof AspectRatioPrimitive.Root;

export { AspectRatio };
