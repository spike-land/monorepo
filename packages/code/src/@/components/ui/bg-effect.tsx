import React, { useState, useEffect } from "react";
import { css, keyframes } from "@emotion/react";
import { cn } from "@/lib/utils";
import { useDarkMode } from "@/hooks/use-dark-mode";


const fadeIn = keyframes`
from { opacity: 0; transform: translateY(20px); }
to { opacity: 1; transform: translateY(0); }
`;

const pulse = keyframes`
0%, 100% { opacity: 1; transform: scale(1); }
50% { opacity: 0.8; transform: scale(0.98); }
`;
const rotate = keyframes`
from { transform: rotate(0deg); }
to { transform: rotate(360deg); }
`;
const cornerRotate = keyframes`
from { transform: rotate(0deg); }
to { transform: rotate(360deg); }
`;
const moveToCorner = (point: string) => {
  const [x, y] = point.split(",").map(Number);
  return `translate(${x - 50}px, ${y - 50}px)`;
};
export const BackgroundEffect: React.FC<{children: React.ReactElement}> = ({children}) => {

  const [mousePosition, setMousePosition] = useState({ x: 0, y: 0 });
  const { isDarkMode } = useDarkMode();
  useEffect(() => {
    const handleMouseMove = (e: MouseEvent) => {
      setMousePosition({ x: e.clientX, y: e.clientY });
    };
    window.addEventListener("mousemove", handleMouseMove);
    return () => {
      window.removeEventListener("mousemove", handleMouseMove);
    };
  }, []);
  return (
    <div
      className={cn(
        "absolute inset-0 overflow-hidden",
        isDarkMode ? "bg-black" : "bg-white",
      )}>
      <div
        className={cn(
          "absolute inset-0 opacity-50 z-10",
          isDarkMode ? "bg-black" : "bg-white",
        )}></div>
      <svg
        css={css`
                 animation: ${pulse} 4s ease-in-out infinite;
                 filter: blur(8px);
                 transform: translate(${(mousePosition.x / window.innerWidth - 0.5) * 20}px, ${(mousePosition.y / window.innerHeight - 0.5) * 20}px);
                 transition: transform 0.1s ease-out;
               `}
        className={cn("absolute inset-0 w-full h-full z-0")}
        xmlns='http://www.w3.org/2000/svg'
        viewBox='0 0 100 100'
        preserveAspectRatio='xMidYMid slice'>
        <defs>
          {["0,0", "100,0", "0,100", "100,100", "50,0", "100,50", "50,100", "0,50"].map(
            (point, index) => (
              <radialGradient
                key={index}
                id={`grad${index}`}
                cx={point.split(",")[0]}
                cy={point.split(",")[1]}
                r='100%'
                gradientUnits='userSpaceOnUse'>
                <stop
                  offset='0%'
                  stopColor={isDarkMode ? "#ff00ff" : "#ff8080"}
                  css={css`
                        animation: ${pulse} ${3 + index}s ease-in-out infinite alternate;
                      `}
                />
                <stop
                  offset='25%'
                  stopColor={isDarkMode ? "#8000ff" : "#ff8000"}
                />
                <stop
                  offset='50%'
                  stopColor={isDarkMode ? "#0080ff" : "#80ff00"}
                />
                <stop
                  offset='75%'
                  stopColor={isDarkMode ? "#00ffff" : "#0080ff"}
                  css={css`
                           animation: ${pulse} ${4 + index}s ease-in-out infinite alternate-reverse;
                         `}
                />
                <stop
                  offset='100%'
                  stopColor={isDarkMode ? "#000000" : "#ffffff"}
                  stopOpacity='0'
                />
              </radialGradient>
            ),
          )}
        </defs>
        {["0,0", "100,0", "0,100", "100,100", "50,0", "100,50", "50,100", "0,50"].map(
          (point, index) => (
            <g
              key={index}
              css={css`
                       transform-origin: 50% 50%;
                       animation: ${cornerRotate} ${20 + index * 3}s linear infinite${index % 2 === 0 ? "" : " reverse"};
                       transform: ${moveToCorner(point)} rotate(${(mousePosition.x / window.innerWidth - 0.5) * 10}deg);
                       transition: transform 0.1s ease-out;
                     `}>
              {[...Array(72)].map((_, i) => (
                <path
                  key={i}
                  d={`M${point} L${parseFloat(point.split(",")[0]) + 60 * Math.cos((i * Math.PI) / 36)},${parseFloat(point.split(",")[1]) + 60 * Math.sin((i * Math.PI) / 36)}`}
                  stroke={`url(#grad${index % 4})`}
                  strokeWidth='0.3'
                  fill='none'
                />
              ))}
            </g>
          ),
        )}
      </svg>
      {children}
    </div>
  );
};
