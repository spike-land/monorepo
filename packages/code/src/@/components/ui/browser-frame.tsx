import { cn } from "@/lib/utils";
import * as React from "react";
import { animated, useSpring } from "react-spring";

interface BrowserFrameProps extends React.HTMLAttributes<HTMLDivElement> {
  name: string;
  description: string;
  isHovered: boolean;
}

const BrowserFrame = React.forwardRef<HTMLDivElement, BrowserFrameProps>(
  ({ children, name, description, isHovered, className, ...props }, ref) => {
    const animation = useSpring({
      transform: isHovered ? "translateY(0%)" : "translateY(100%)",
      opacity: isHovered ? 1 : 0,
    });

    return (
      <div
        ref={ref}
        className={cn(
          "rounded-lg overflow-hidden shadow-lg relative transition-all duration-500 w-full h-0 pb-[75%] bg-white bg-opacity-70 hover:bg-opacity-100",
          className,
        )}
        {...props}
      >
        <div className="bg-gray-100 p-2 flex items-center space-x-2">
          <div className="flex space-x-1">
            {["red", "yellow", "green"].map((color) => (
              <div key={color} className={`w-3 h-3 rounded-full bg-${color}-400`} />
            ))}
          </div>
          <div className="flex-grow bg-white rounded px-2 py-1 text-sm text-gray-600">
            {name}
          </div>
        </div>
        <div className="absolute top-[40px] left-0 right-0 bottom-0 bg-white overflow-hidden">
          <div className="w-[calc(100%+4px)] h-[calc(100%+4px)] -m-[2px]">{children}</div>
        </div>
        <animated.div
          style={animation}
          className="absolute bottom-0 left-0 right-0 bg-black bg-opacity-80 text-white text-sm p-3 text-center rounded-b-lg"
        >
          {description}
        </animated.div>
      </div>
    );
  },
);

BrowserFrame.displayName = "BrowserFrame";

export { BrowserFrame };
