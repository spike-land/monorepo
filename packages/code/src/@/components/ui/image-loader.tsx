import React from "react";

type RESOLUTION = {
  "9:21": [640, 1536];
  "9:16": [768, 1344];
  "16:9": [896, 1584];
  "16:10": [896, 1408];
  "5:4": [1088, 896];
  "4:5": [896, 1088];
  "2:3": [832, 1216];
  "3:2": [1216, 832];
  "1:1": [1024, 1024];
};

interface ImageLoaderProps {
  cfg?: number;
  steps?: number;
  prompt: string;
  aspect_ratio?: keyof RESOLUTION;
  output_format?: "webp" | "png" | "jpeg";
  output_quality?: number;
  negative_prompt?: string;
  prompt_strength?: number;
  className?: string;
}

const DEFAULT_PROPS: Required<ImageLoaderProps> = {
  cfg: 3.5,
  steps: 28,
  prompt: "A web app for AI development. spike.land",
  aspect_ratio: "16:9",
  output_format: "webp",
  output_quality: 90,
  negative_prompt: "",
  prompt_strength: 0.85,
  className: "",
};

export const ImageLoader: React.FC<ImageLoaderProps> = (props) => {
  const params = React.useMemo(() => {
    const params = new URLSearchParams();
    (Object.keys(DEFAULT_PROPS) as Array<keyof ImageLoaderProps>).forEach(
      (key) => {
        if (
          key && props[key] && key !== "className" &&
          props[key] !== DEFAULT_PROPS[key]
        ) {
          params.append(key, String(props[key]));
        }
      },
    );
    return params.toString();
  }, [props]);

  const containerClassName = props.className || "w-full max-w-2xl mx-auto";

  return (
    <img
      src={`/replicate/${btoa(params)}.${
        props.output_format || DEFAULT_PROPS.output_format
      }`}
      alt={props.prompt || DEFAULT_PROPS.prompt}
      className={containerClassName}
    />
  );
};

export default () => <ImageLoader prompt="A nice day" />;
