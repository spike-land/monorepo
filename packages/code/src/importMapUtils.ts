export const oo: { [key: string]: string } = {
  "@emotion/react/jsx-runtime": "/emotionJsxRuntime.mjs",
  "react/jsx-runtime": "/jsx.mjs",
  "react-dom/server": "/reactDomServer.mjs",
  "react-dom/client": "/reactDomClient.mjs",
  "@emotion/react": "/emotion.mjs",
  "react": "/reactMod.mjs",
  "framer-motion": "/motion.mjs",
  "react-dom": "/reactDom.mjs",
  "foo-bar": "/fooBar.mjs",
  "recharts": "/recharts.mjs",
};

export const importMap = { imports: oo };

export function importMapReplace(code: string, origin: string): string {
  // Check if the code has already been processed
  if (code.startsWith("/** importMapReplace */")) {
    return code;
  }

  const importRegex = /^([ \t]*import\s+(?:[\w*{}\n\r\t, ]+)\s+from\s+)["'](.*[@\w_-]+)["']\s*;?\s*(\/\/.*)?\s*$/gm;
  const dynamicImportRegex = /import\((["'`])(.+?)\1\)/g;
  const exportRegex = /^([ \t]*export\s+(?:{\s*[\w\s,]+\s*}\s+from\s+)?)["'](.*[@\w_-]+)["']\s*;?\s*(\/\/.*)?\s*$/gm;

  function replaceImport(match: string, importStatement: string, path: string, inlineComment: string = ""): string {
    let newPath = path;

    if (oo[path]) {
      newPath = `${origin}${oo[path]}`;
    } else if (!path.startsWith("http") && !path.startsWith("./") && !path.startsWith("../")) {
      if (path.startsWith("@/")) {
        newPath = `${origin}/@/${path.slice(2)}`;
      } else {
        newPath = `${origin}/${path}`;
      }
      if (!newPath.includes("?bundle")) {
        newPath += path === "axios" ? "?bundle" : ".mjs";
      }
    } else if (path.startsWith("./") || path.startsWith("../")) {
      newPath = `${origin}/live/${path.replace(/^\.\//, "")}`;
      if (!newPath.endsWith(".js") && !newPath.endsWith(".mjs")) {
        newPath += "/index.js";
      }
    } else {
      return match;
    }

    // Extract specific exports if present
    const exportsMatch = importStatement.match(
      /{\s*([\w\s,]+(?:\s+as\s+[\w]+)?(?:\s*,\s*[\w]+(?:\s+as\s+[\w]+)?)*)\s*}/,
    );
    if (exportsMatch && !oo[path] && !path.startsWith("@/")) {
      const specificExports = exportsMatch[1].replace(/\s+as\s+/g, " as ").trim();
      newPath += newPath.includes("?") ? "&" : "?";
      newPath += `exports=${encodeURIComponent(specificExports)}`;
    }

    return `${importStatement}"${newPath}";${inlineComment ? " " + inlineComment : ""}`;
  }

  function replaceDynamicImport(match: string, quote: string, path: string): string {
    if (path.includes("${")) {
      return match;
    }
    if (oo[path]) {
      return `import(${quote}${origin}${oo[path]}${quote})`;
    }
    if (!path.startsWith("http") && !path.startsWith("./") && !path.startsWith("../")) {
      if (path.startsWith("@/")) {
        return `import(${quote}${origin}/@/${path.slice(2)}?bundle${quote})`;
      }
      return `import(${quote}${origin}/${path.includes("*") ? path : "*" + path + "?bundle"}${quote})`;
    }
    return match;
  }

  function replaceExport(match: string, exportStatement: string, path: string, inlineComment: string = ""): string {
    let newPath = path;

    if (oo[path]) {
      newPath = `${origin}${oo[path]}`;
    } else if (!path.startsWith("http") && !path.startsWith("./") && !path.startsWith("../")) {
      if (path.startsWith("@/")) {
        newPath = `${origin}/@/${path.slice(2)}`;
      } else {
        newPath = `${origin}/${path}`;
      }
      newPath += ".mjs";
    } else if (path.startsWith("./") || path.startsWith("../")) {
      newPath = `${origin}/live/${path.replace(/^\.\//, "")}`;
      if (!newPath.endsWith(".js") && !newPath.endsWith(".mjs")) {
        newPath += "/index.js";
      }
    } else {
      return match;
    }

    return `${exportStatement}"${newPath}";${inlineComment ? " " + inlineComment : ""}`;
  }

  code = code.replace(importRegex, replaceImport);
  code = code.replace(dynamicImportRegex, replaceDynamicImport);
  code = code.replace(exportRegex, replaceExport);

  return `/** importMapReplace */\n${code.trim()}`;
}
