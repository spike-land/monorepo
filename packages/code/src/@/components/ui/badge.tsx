import { badgeVariants } from "@/components/ui/badge-variants";
import { cn } from "@/lib/utils";
import { type VariantProps } from "class-variance-authority";
import type { HTMLAttributes } from "react";

export interface BadgeProps
  extends HTMLAttributes<HTMLDivElement>, VariantProps<typeof badgeVariants>
{}

function Badge({ className, variant, ...props }: BadgeProps) {
  return <div className={cn(badgeVariants({ variant }), className)} {...props} />;
}

export { Badge };
