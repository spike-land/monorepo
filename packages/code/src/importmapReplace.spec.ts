import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { importMapReplace } from "./importMapUtils"; // Updated import

describe("importMapReplace", () => {
  const origin = "http://localhost:3000";

  beforeAll(() => {
    vi.stubGlobal(
      "fetch",
      vi.fn().mockImplementation(() =>
        Promise.resolve({
          ok: true,
          json: () =>
            Promise.resolve({
              "name": "some-module",
              "version": "1.0.0",
              "main": "index.js",
              "browser": "browser.js",
              "module": "module.js",
              "types": "index.d.ts",
              "files": [
                "index.js",
                "browser.js",
                "module.js",
                "index.d.ts",
              ],
            }),
        })
      ),
    );
  });

  afterAll(() => {
    vi.unstubAllGlobals();
  });

  it("should replace top-level imports", async () => {
    const code = "import React from \"react\";";
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace top-level imports with dot in the module name", async () => {
    const code = "import React from \"react.gl\";";
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should handle relative imports", async () => {
    const code = "import React from \"./box\";";
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace shadcn-ui imports", async () => {
    const code = `
    import { css } from "@emotion/react";
    import $ from "react";
    import { Button } from "@/components/ui/button";

    import { css  as $kdkd } from "@emotion/react";

    import { useEffect, useState, useRef } from "react";
    import type { FC } from "react";
    import QRTerminal from "qrcode-terminal";
    `;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should handle multiple imports from different modules", async () => {
    const code = `
    import { css } from "@emotion/react";
    import $ from "react";

    import { css  as $kdkd } from "@emotion/react";

    import { useEffect, useState, useRef } from "react";
    import type { FC } from "react";
    import QRTerminal from "qrcode-terminal";
    `;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace top-level exports, except the ones in importmap", async () => {
    const code = `import { useState } from "react";`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace top-level default exports", async () => {
    const code = `export { default } from "react";`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace dynamic imports", async () => {
    const code = `const React = import("react");`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should handle dynamic imports with template literals", async () => {
    const code = "const dynamic = 'dd'; const React = import(`${dynamic}dom`);";
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
    expect(result).includes("React = import(`${dynamic}dom`);");
  });

  it("should replace shadcn-ui dynamic imports", async () => {
    const code = "import React from \"@/components/ui/alert\";";
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should ignore relative URLs", async () => {
    const code = `import MyComponent from "./MyComponent";`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should ignore absolute URLs", async () => {
    const code = "import MyComponent from \"http://example.com/MyComponent.js\";";
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should apply custom mappings", async () => {
    const code = "import CustomPackage from \"custom-package\";";
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace dynamic imports with await", async () => {
    const code = "const module = await import(\"some-module\");";
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace dynamic imports inside an async function", async () => {
    const code = `
    import{Buffer as De}from"/v125/buffer@6.0.3/es2024/buffer.bundle.mjs"
    let { Mp3Encoder: t } = await import("lamejs");
    import{timeDay as _,timeSunday as bt,timeMonday as b,timeThursday as w,timeYear as S,utcDay as $,utcSunday as Vt,utcMonday as V,utcThursday as p,utcYear as F}from"d3-time";
      async function loadModule() {
        let{default:c}=await import("axios");
        let{default:d}=await import("https://unpkg.com/axios/dist/axios.min.js");
        
        const module = await import("some-module");
        // do something with module
      }
    `;

    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should process JSX-related imports", async () => {
    const code =
      `eturn d.className=f,a&&(d.ref=a),O.createElement(O.Fragment,null,O.createElement(On,{cache:t,serialized:m,isStringTag:typeof c=="string"}),O.createElement(c,d))}),dr=Rn;import"react";var va=Y(Yr());var xa=Z.Fragment;function Sa(e,t,a){return me.call(t,"css")?Z.jsx(dr,pr(e,t),a):Z.jsx(e,t,a)}function Ea(e,t,a){return me.call(t,"css")?Z.jsxs(dr,pr(e,t),a):Z.jsxs(e,t,a)}";`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace the live URLs", async () => {
    const code = `import Box from "/live/Box";`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should not modify non-import statements", async () => {
    const code = ` eturn e[t] === r ||
              (e[t] = r,
                (t === "prop" || t === "value" || t === "name" ||
                  t === "params" || t === "important" || t === "text") &&
                e.markDirty()),
              !0;`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace dynamic imports at the end of the file", async () => {
    const code = `
    import{Buffer as De}from"/v125/buffer@6.0.3/es2024/buffer.bundle.mjs"
    let { Mp3Encoder: t } = await import("lamejs");
    import{timeDay as _,timeSunday as bt,timeMonday as b,timeThursday as w,timeYear as S,utcDay as $,utcSunday as Vt,utcMonday as V,utcThursday as p,utcYear as F}from"d3-time";
      async function loadModule() {
        let{default:c}=await import("axios");
        let{default:d}=await import("https://unpkg.com/axios/dist/axios.min.js");
        
        const module = await import("some-module");
        // do something with module
      }
        import"react/jsx-runtime";
        import"react";
        async function y(s){await s(m)}
        export{n as Particles,n as default,y as initParticlesEngine};
    `;

    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should replace imports from react-spring", async () => {
    const code = `
    import { useSpring, animated } from "react-spring"; // Importing react-spring
    `;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should handle specific named exports", async () => {
    const code = `import { prop, prop2 } from "foo";`;
    const result = importMapReplace(code, origin);
    expect(result).toContain(`"${origin}/*foo?bundle=true&exports=prop,prop2"`);
    expect(result).toMatchSnapshot();
  });

  it("should handle specific exports from tslib", async () => {
    const code = `import { __await, __rest } from "tslib";`;
    const result = importMapReplace(code, origin);
    expect(result).toContain(`"${origin}/*tslib?bundle=true&exports=__await,__rest"`);
    expect(result).toMatchSnapshot();
  });

  it("should handle specific named imports with aliases", async () => {
    const code = `import { __await as aw, __rest  as restNow} from "tslib";`;
    const result = importMapReplace(code, origin);
    expect(result).toContain(`"${origin}/*tslib?bundle=true&exports=__await,__rest"`);
    expect(result).toMatchSnapshot();
  });

  it("should handle import statements with multiple named imports", async () => {
    const code = `import { useState, useEffect, useCallback } from "react";`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should handle import statements with both default and named imports", async () => {
    const code = `import React, { useState, useEffect } from "react";`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should handle import statements with namespace imports", async () => {
    const code = `import * as React from "react";`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should handle import statements with default imports and aliases", async () => {
    const code = `import ReactDOM from "react-dom/client";`;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it.only("should handle multiple import statements", async () => {
    const code = `
      import React from "react";
      import { useState } from "react";
      import axios from "axios";
    `;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  // New test cases
  it("should handle imports with file extensions", async () => {
    const code = `
      import MyModule from "./myModule.js";
      import MyTypeScriptModule from "./myModule.ts";
      import MyJSXModule from "./myModule.jsx";
      import MyTSXModule from "./myModule.tsx";
    `;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should handle imports starting with @/", async () => {
    const code = `
      import MyComponent from "@/components/MyComponent";
      import MyUtil from "@/utils/MyUtil";
    `;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });

  it("should not modify code that already contains importMapReplace", async () => {
    const code = `
      /** importMapReplace */
      import React from "react";
    `;
    const result = importMapReplace(code, origin);
    expect(result).toBe(code);
  });

  it("should handle imports with query parameters", async () => {
    const code = `
      import MyModule from "my-module?param1=value1&param2=value2";
    `;
    const result = importMapReplace(code, origin);
    expect(result).toMatchSnapshot();
  });
});
