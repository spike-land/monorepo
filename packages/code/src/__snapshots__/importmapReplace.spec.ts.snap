// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`importMapReplace > should apply custom mappings 1`] = `
"
  /** importMapReplace */
  import CustomPackage from "http://localhost:3000/*custom-package?bundle";"
`;

exports[`importMapReplace > should handle dynamic imports with template literals 1`] = `
"
  /** importMapReplace */
  const dynamic = 'dd'; const React = import(\`\${dynamic}dom\`);"
`;

exports[`importMapReplace > should handle import statements with both default and named imports 1`] = `
"
  /** importMapReplace */
  import React, { useState, useEffect } from "http://localhost:3000/reactMod.mjs";"
`;

exports[`importMapReplace > should handle import statements with default imports and aliases 1`] = `
"
  /** importMapReplace */
  import ReactDOM from "http://localhost:3000/reactDomClient.mjs";"
`;

exports[`importMapReplace > should handle import statements with multiple named imports 1`] = `
"
  /** importMapReplace */
  import { useState, useEffect, useCallback } from "http://localhost:3000/reactMod.mjs";"
`;

exports[`importMapReplace > should handle import statements with namespace imports 1`] = `
"
  /** importMapReplace */
  import * as React from "http://localhost:3000/reactMod.mjs";"
`;

exports[`importMapReplace > should handle imports starting with @/ 1`] = `
"
  /** importMapReplace */
  
      import MyComponent from "http://localhost:3000/@/components/MyComponent.mjs";
      import MyUtil from "http://localhost:3000/@/utils/MyUtil.mjs";
    "
`;

exports[`importMapReplace > should handle imports with file extensions 1`] = `
"
  /** importMapReplace */
  
      import MyModule from "./myModule.js";
      import MyTypeScriptModule from "./myModule.ts";
      import MyJSXModule from "./myModule.jsx";
      import MyTSXModule from "./myModule.tsx";
    "
`;

exports[`importMapReplace > should handle imports with query parameters 1`] = `
"
  /** importMapReplace */
  
      import MyModule from "http://localhost:3000/*my-module?param1=value1&param2=value2?bundle";
    "
`;

exports[`importMapReplace > should handle multiple import statements 1`] = `
"/** importMapReplace */
import React from "http://localhost:3000/reactMod.mjs";
      import { useState } from "http://localhost:3000/reactMod.mjs";
      import axios from "http://localhost:3000/axios?bundle";"
`;

exports[`importMapReplace > should handle multiple imports from different modules 1`] = `
"
  /** importMapReplace */
  
    import { css } from "http://localhost:3000/emotion.mjs";
    import $ from "http://localhost:3000/reactMod.mjs";

    import { css  as $kdkd } from "http://localhost:3000/emotion.mjs";

    import { useEffect, useState, useRef } from "http://localhost:3000/reactMod.mjs";
    import type { FC } from "http://localhost:3000/reactMod.mjs";
    import QRTerminal from "http://localhost:3000/*qrcode-terminal?bundle";
    "
`;

exports[`importMapReplace > should handle relative imports 1`] = `
"
  /** importMapReplace */
  import React from "http://localhost:3000/live/box/index.js";"
`;

exports[`importMapReplace > should handle specific exports from tslib 1`] = `
"
  /** importMapReplace */
  import { __await, __rest } from "http://localhost:3000/*tslib?bundle=true&exports=__await,__rest";"
`;

exports[`importMapReplace > should handle specific named exports 1`] = `
"
  /** importMapReplace */
  import { prop, prop2 } from "http://localhost:3000/*foo?bundle=true&exports=prop,prop2";"
`;

exports[`importMapReplace > should handle specific named imports with aliases 1`] = `
"
  /** importMapReplace */
  import { __await as aw, __rest  as restNow} from "http://localhost:3000/*tslib?bundle=true&exports=__await,__rest";"
`;

exports[`importMapReplace > should ignore absolute URLs 1`] = `
"
  /** importMapReplace */
  import MyComponent from "http://example.com/MyComponent.js";"
`;

exports[`importMapReplace > should ignore relative URLs 1`] = `
"
  /** importMapReplace */
  import MyComponent from "http://localhost:3000/live/MyComponent/index.js";"
`;

exports[`importMapReplace > should not modify non-import statements 1`] = `
"
  /** importMapReplace */
   eturn e[t] === r ||
              (e[t] = r,
                (t === "prop" || t === "value" || t === "name" ||
                  t === "params" || t === "important" || t === "text") &&
                e.markDirty()),
              !0;"
`;

exports[`importMapReplace > should process JSX-related imports 1`] = `
"
  /** importMapReplace */
  eturn d.className=f,a&&(d.ref=a),O.createElement(O.Fragment,null,O.createElement(On,{cache:t,serialized:m,isStringTag:typeof c=="string"}),O.createElement(c,d))}),dr=Rn;import"http://localhost:3000/reactMod.mjs";var va=Y(Yr());var xa=Z.Fragment;function Sa(e,t,a){return me.call(t,"css")?Z.jsx(dr,pr(e,t),a):Z.jsx(e,t,a)}function Ea(e,t,a){return me.call(t,"css")?Z.jsxs(dr,pr(e,t),a):Z.jsxs(e,t,a)}";"
`;

exports[`importMapReplace > should replace dynamic imports 1`] = `
"
  /** importMapReplace */
  const React = import("http://localhost:3000/reactMod.mjs");"
`;

exports[`importMapReplace > should replace dynamic imports at the end of the file 1`] = `
"
  /** importMapReplace */
  
    import{Buffer as De}from"http://localhost:3000/v125/buffer@6.0.3/es2024/buffer.bundle.mjs"
    let { Mp3Encoder: t } = await import("http://localhost:3000/*lamejs?bundle");
    import{timeDay as _,timeSunday as bt,timeMonday as b,timeThursday as w,timeYear as S,utcDay as $,utcSunday as Vt,utcMonday as V,utcThursday as p,utcYear as F}from"http://localhost:3000/*d3-time?bundle=true&exports=timeDay,timeSunday,timeMonday,timeThursday,timeYear,utcDay,utcSunday,utcMonday,utcThursday,utcYear";
      async function loadModule() {
        let{default:c}=await import("http://localhost:3000/*axios?bundle");
        let{default:d}=await import("https://unpkg.com/axios/dist/axios.min.js");
        
        const module = await import("http://localhost:3000/*some-module?bundle");
        // do something with module
      }
        import"http://localhost:3000/jsx.mjs";
        import"http://localhost:3000/reactMod.mjs";
        async function y(s){await s(m)}
        export{n as Particles,n as default,y as initParticlesEngine};
    "
`;

exports[`importMapReplace > should replace dynamic imports inside an async function 1`] = `
"
  /** importMapReplace */
  
    import{Buffer as De}from"http://localhost:3000/v125/buffer@6.0.3/es2024/buffer.bundle.mjs"
    let { Mp3Encoder: t } = await import("http://localhost:3000/*lamejs?bundle");
    import{timeDay as _,timeSunday as bt,timeMonday as b,timeThursday as w,timeYear as S,utcDay as $,utcSunday as Vt,utcMonday as V,utcThursday as p,utcYear as F}from"http://localhost:3000/*d3-time?bundle=true&exports=timeDay,timeSunday,timeMonday,timeThursday,timeYear,utcDay,utcSunday,utcMonday,utcThursday,utcYear";
      async function loadModule() {
        let{default:c}=await import("http://localhost:3000/*axios?bundle");
        let{default:d}=await import("https://unpkg.com/axios/dist/axios.min.js");
        
        const module = await import("http://localhost:3000/*some-module?bundle");
        // do something with module
      }
    "
`;

exports[`importMapReplace > should replace dynamic imports with await 1`] = `
"
  /** importMapReplace */
  const module = await import("http://localhost:3000/*some-module?bundle");"
`;

exports[`importMapReplace > should replace imports from react-spring 1`] = `
"
  /** importMapReplace */
  
    import { useSpring, animated } from "http://localhost:3000/*react-spring?bundle=true&exports=useSpring,animated"; // Importing react-spring
    "
`;

exports[`importMapReplace > should replace shadcn-ui dynamic imports 1`] = `
"
  /** importMapReplace */
  import React from "http://localhost:3000/@/components/ui/alert.mjs";"
`;

exports[`importMapReplace > should replace shadcn-ui imports 1`] = `
"
  /** importMapReplace */
  
    import { css } from "http://localhost:3000/emotion.mjs";
    import $ from "http://localhost:3000/reactMod.mjs";
    import { Button } from "http://localhost:3000/@/components/ui/button.mjs";

    import { css  as $kdkd } from "http://localhost:3000/emotion.mjs";

    import { useEffect, useState, useRef } from "http://localhost:3000/reactMod.mjs";
    import type { FC } from "http://localhost:3000/reactMod.mjs";
    import QRTerminal from "http://localhost:3000/*qrcode-terminal?bundle";
    "
`;

exports[`importMapReplace > should replace the live URLs 1`] = `
"
  /** importMapReplace */
  import Box from "http://localhost:3000/live/Box";"
`;

exports[`importMapReplace > should replace top-level default exports 1`] = `
"
  /** importMapReplace */
  export { default } from "http://localhost:3000/reactMod.mjs";"
`;

exports[`importMapReplace > should replace top-level exports, except the ones in importmap 1`] = `
"
  /** importMapReplace */
  import { useState } from "http://localhost:3000/reactMod.mjs";"
`;

exports[`importMapReplace > should replace top-level imports 1`] = `
"
  /** importMapReplace */
  import React from "http://localhost:3000/reactMod.mjs";"
`;

exports[`importMapReplace > should replace top-level imports with dot in the module name 1`] = `
"
  /** importMapReplace */
  import React from "http://localhost:3000/*react.gl?bundle";"
`;
